### DISCUSSION TITLE
how can i use for(var i:x..y) in C#  8.0 ranges

### DISCUSSION THREAD

**Discussion by sgf** [dotnet/csharplang#3939] [General] [answered]
like this:
```C#
for(var i:1..10)
{
 do some thing with i
}
//or
for(var i=1..10)
{
 do some thing with i
}
//or
for(var i in 1..10)
{
 do some thing with i
}
//or
foreach(var i in 1..10)
{
 do some thing with i
}
```

**how can i do one of any like that ?
or 
c#‘s ranges are currently unable to do so.**

**Comment by YairHalberstadt** [comment, marked as answer]
In c# 9.0 you can reference https://www.nuget.org/packages/RangeForeach/0.0.1 which uses extension foreach, allowing you to do:
```csharp
foreach(var i in 1..10)
{
 do some thing with i
}
```

**Reply by sgf** [reply]
thanks,thats a good idea

**Reply by DavidArno** [reply]
It's disappointing that you've had to release that as a nuget package rather than it being added to the 5.0 CLR via a pull request. But it's still good that you wrote it.

**Reply by 333fred** [reply]
For reference, the reason you have to do that is because ranges have an exclusive end. That will foreach over the numbers 1 through 9, not 1 through 10 (unless Yair changed the behavior since we last talked about it).

**Reply by jnm2** [reply]
@DavidArno The problem with adding to the BCL is that people will disagree about how it should work (exclusivity, reverse iteration) and it's not a perfect abstraction because either or both ends of a range can be from-end which makes no sense to loop over.

**Reply by jnm2** [reply]
I'd almost rather it wasn't in the BCL and that the compiler lowered it directly, allowing only iteration over the subset range literals that aren't from-end. You'd still have to decide about exclusivity and reverse iteration.

**Reply by sgf** [reply]
maybe could be let the:
1..10 from 1 to 9 (C#8.0 compatible)
1...10 from 1 to 10;

**Reply by sgf** [reply]
This syntax will bring convenience to developers and is more intuitive than [for loop] -"for(int i=0;i<xx;i++)". 
So it would be great if could adding BLC.

**Reply by DavidArno** [reply]
The different between `1 .. 10` and `1 ... 10` is too small in my view and easily missed when reading the code.

I think there was discussion around using `1 to 10` as syntax for inclusive ranges but I think that died a death.

**Reply by JustNrik** [reply]
1..10 should go from 1 to 9, that's how range works and that's how it's implemented in many other languages like python.

why would 1..10 be ambiguous on foreach but not in an array[1..10]? there's no ambiguity, the feature is clear.

I would also agree for inclusive syntax, actually, we should have inclusive range like swift, like `1..=10`? kinda ugly, I can't think of a possible syntax that fits c# atm
